src/components/CreateAgent/Persona/index.tsx

```diff
@@ -53,45 +61,108 @@ function Persona({ wizard, setCurrentNav, setWizardId, isEdit }: PersonaProps) {
     } as WizardVisibility;
     if (userId === undefined) return;
     if (wizardName === null) return;
+
     if (isEdit) {
-      updateWizard(
-        {
-          wizardId: wizard.id,
-          updatedWizardDetails: {
-            ...values,
-            name: wizardName ? wizardName : wizard.name,
-            visibility,
+      if(wizard.avatar.slice(0,11) === "default_img") {
+        try {
+          const uploadSuccess = await handleUploadCustomImage(values);
+          if (uploadSuccess) handleAddWizard(values, userId, visibility);
+        } catch (error) {
+          console.error("Upload failed", error);
+        }
+      }
+      else {
+        deleteCustomImage(wizard.avatar, {
+          onSuccess: async () => {
+            try {
+              const uploadSuccess = await handleUploadCustomImage(values);
+              if (uploadSuccess) handleUpdateWizard(values, visibility);
+            } catch (error) {
+              console.error("Upload failed", error);
+            }
           },
-        },
-        {
-          onSuccess: () =>
-            queryClient.invalidateQueries({ queryKey: [wizard.id] }),
           onError: error => {
             console.error(error);
-            toast.error(error.message);
           },
-        }
-      );
+        })
+      }
     } else {
-      const payload: WizardDetails = {
-        ...values,
-        id: uuidv4(),
-        userId,
-        name: wizardName,
-        visibility,
-        summary: [],
-        isPublished: false,
-      };
-      addWizard(payload, {
-        onSuccess: () => {
-          setWizardId(payload.id);
-          setCurrentNav("knowledge");
-        },
-        onError: error => toast.error(error.message),
+      try {
+        const uploadSuccess = await handleUploadCustomImage(values);
+        if (uploadSuccess) handleAddWizard(values, userId, visibility);
+      } catch (error) {
+          console.error("Upload failed", error);
+      }
+    }
+  };
+
+  const handleUploadCustomImage = async (values: PersonaValues) => {
+    if (customImageNameRef.current !== '') {
+      return new Promise((resolve, reject) => {
+        uploadCustomImage({
+          fileName: customImageNameRef.current,
+          base64Image: values.avatar
+        }, {
+          onSuccess: (response) => {
+            if ('Ok' in response) {
+              const id = response.Ok;
+              values.avatar = id;
+              resolve(true);
+            } else {
+              reject(false);
+            }
+          },
+          onError: error => {
+            console.error(error);
+            reject(false);
+          },
+        });
       });
+    } else {
+      return Promise.resolve(true);
```

Comment:
why is this a promise?

Reply:
I had tried async await/.then, the functions werent getting executed sequentially (upload image -> get key and then add wizard).
They both kept getting executed simultaneously and add wizard always returned way before upload image.

We can check this out if you want to use async await

Reply:
instead of this you can do it something like 
```
if (customImageNameRef.current === "") {
  // throw error or toast 
}

uploadCustomImage(
  {
    fileName: customImageNameRef.current,
    base64Image: values.avatar,
  },
  {
    onSuccess: (response) => {
      if ("Ok" in response) {
        const id = response.Ok;
        values.avatar = id;
        const payload: WizardDetails = {
          ...values,
          id: uuidv4(),
          userId,
          name: wizardName,
          visibility,
          summary: [],
          isPublished: false,
        };
        addWizard(payload, {
          onSuccess: () => {
            setWizardId(payload.id);
            setCurrentNav("knowledge");
          },
          onError: error => toast.error(error.message),
        });
      } else {
        // reject(false);
        // throw error or toast 
      }
    },
    onError: (error) => {
      console.error(error);
    },
  }
);

```

Reply:
I had rewritten in a similar way today, this makes sense,
Have also organised the edge cases better in the new commit along with all the cleanup

https://github.com/elna-ai/ELNA-DApp/pull/210/commits/68f356d055974ca52c48a525d753608fc884f1f4